core/src/test/java/io/grpc/internal/ServiceConfigErrorHandlingTest.java

```diff
@@ -0,0 +1,736 @@
+
+    createChannel();
+
+    ArgumentCaptor<ResolvedAddresses> resultCaptor =
+        ArgumentCaptor.forClass(ResolvedAddresses.class);
+    verify(mockLoadBalancer).handleResolvedAddresses(resultCaptor.capture());
+    ResolvedAddresses resolvedAddresses = resultCaptor.getValue();
+    assertThat(resolvedAddresses.getAddresses()).containsExactly(addressGroup);
+    assertThat(getLbPolicyConfig(resolvedAddresses)).isNotNull();
+    assertThat(resolvedAddresses.getAttributes().get(GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG))
+        .isEqualTo(rawServiceConfig);
+    verify(mockLoadBalancer, never()).handleNameResolutionError(any(Status.class));
+
+    assertThat(channel.getState(false)).isNotEqualTo(ConnectivityState.TRANSIENT_FAILURE);
+
+    reset(mockLoadBalancer);
+    nameResolverFactory.nextConfigOrError.set(null);
+    Status error = Status.UNKNOWN.withDescription("not known");
+    nameResolverFactory.nextConfigOrError.set(ConfigOrError.fromError(error));
+
+    // 2nd resolution
+    nameResolverFactory.allResolved();
+
+    verify(mockLoadBalancer).handleResolvedAddresses(resultCaptor.capture());
+    ResolvedAddresses newResolvedAddress = resultCaptor.getValue();
+    assertThat(getLbPolicyConfig(resolvedAddresses)).isNotNull();
+    // should use previous service config because new service config is invalid.
+    assertThat(newResolvedAddress.getAttributes()).isNotEqualTo(Attributes.EMPTY);
+    assertThat(newResolvedAddress.getAttributes().get(GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG))
+        .isEqualTo(rawServiceConfig);
+    verify(mockLoadBalancer, never()).handleNameResolutionError(any(Status.class));
+    assertThat(channel.getState(false)).isEqualTo(ConnectivityState.IDLE);
+  }
+
+  private static final class ChannelBuilder
+      extends AbstractManagedChannelImplBuilder<ChannelBuilder> {
+
+    ChannelBuilder() {
+      super(TARGET);
+    }
+
+    @Override protected ClientTransportFactory buildTransportFactory() {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override protected int getDefaultPort() {
+      return DEFAULT_PORT;
+    }
+  }
+
+  private static final class FakeBackoffPolicyProvider implements BackoffPolicy.Provider {
+    @Override
+    public BackoffPolicy get() {
+      return new BackoffPolicy() {
+        int multiplier = 1;
+
+        @Override
+        public long nextBackoffNanos() {
+          return RECONNECT_BACKOFF_INTERVAL_NANOS * multiplier++;
+        }
+      };
+    }
+  }
+
+  private static final class FakeNameResolverFactory extends NameResolver.Factory {
+    final URI expectedUri;
+    final List<EquivalentAddressGroup> servers;
+    final boolean resolvedAtStart;
+    final Status error;
+    final ArrayList<FakeNameResolver> resolvers = new ArrayList<>();
+    final AtomicReference<ConfigOrError> nextConfigOrError = new AtomicReference<>();
+    final AtomicReference<Map<String, ?>> nextRawServiceConfig = new AtomicReference<>();
```

Comment:
`nextConfigOrError` should go away. Instead you will call `ServiceConfigParser.parseServiceConfig()` with `nextRawServiceConfig`, and use the returned `ConfigOrError` to call `listener.onResult()`. In each individual test you should not call `mockLoadBalancerProvider.parseLoadBalancingConfig()` because it's called by `AutoConfiguredLoadBalancerFactory`.